canary/producer-cpp/src/Canary.cpp

```diff
@@ -0,0 +1,899 @@
+                {
+                    cout << "Failed to put PersistedAckLatency metric data:" << outcome.GetError().GetMessage() << endl;
+                }
+                else
+                {
+                    cout << "Successfully put PersistedAckLatency metric data" << endl;
+                }
+            } else if (pFragmentAck->ackType == FRAGMENT_ACK_TYPE_RECEIVED)
+                {
+                    Aws::CloudWatch::Model::MetricDatum receivedAckLatency_datum;
+                    Aws::CloudWatch::Model::PutMetricDataRequest cwRequest;
+                    cwRequest.SetNamespace("KinesisVideoSDKCanaryCPP");
+
+                    auto currentTimestamp = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
+                    auto receivedAckLatency = (currentTimestamp - timeOfFragmentEndSent); // [milliseconds]
+                    pushMetric("ReceivedAckLatency", receivedAckLatency, Aws::CloudWatch::Model::StandardUnit::Milliseconds, receivedAckLatency_datum, data->pDimension_per_stream, cwRequest);
+                    LOG_DEBUG("Received Ack Latencyy: " << receivedAckLatency);
+                    if (data->pCanaryConfig->useAggMetrics)
+                    {
+                        pushMetric("ReceivedAckLatency", receivedAckLatency, Aws::CloudWatch::Model::StandardUnit::Milliseconds, receivedAckLatency_datum, data->pAggregated_dimension, cwRequest);
+                    }
+
+                    auto outcome = data->pCWclient->PutMetricData(cwRequest);
+                    if (!outcome.IsSuccess())
+                    {
+                        cout << "Failed to put ReceivedAckLatency metric data:" << outcome.GetError().GetMessage() << endl;
+                    }
+                    else
+                    {
+                        cout << "Successfully put ReceivedAckLatency metric data" << endl;
+                    }
+                }
+        } else
+        {
+            cout << "Not sending Ack Latency metric because: timeOfFragmentEndSent < pFragmentAck->timestamp" << endl;
+        }
+    }
+}
+
+}  // namespace video
+}  // namespace kinesis
+}  // namespace amazonaws
+}  // namespace com;
+
+// add frame pts, frame index, original frame size, CRC to beginning of buffer
+VOID addCanaryMetadataToFrameData(PFrame pFrame)
+{
+    PBYTE pCurPtr = pFrame->frameData;
+    putUnalignedInt64BigEndian((PINT64) pCurPtr, pFrame->presentationTs / HUNDREDS_OF_NANOS_IN_A_MILLISECOND);
+    pCurPtr += SIZEOF(UINT64);
+    putUnalignedInt32BigEndian((PINT32) pCurPtr, pFrame->index);
+    pCurPtr += SIZEOF(UINT32);
+    putUnalignedInt32BigEndian((PINT32) pCurPtr, pFrame->size);
+    pCurPtr += SIZEOF(UINT32);
+    putUnalignedInt32BigEndian((PINT32) pCurPtr, COMPUTE_CRC32(pFrame->frameData, pFrame->size));
+}
+
+VOID createCanaryFrameData(PFrame pFrame)
+{
+    UINT32 i;
+
+    for (i = CANARY_METADATA_SIZE; i < pFrame->size; i++) {
+        pFrame->frameData[i] = RAND();
```

Comment:
Why this? We dont need random data because we have our actual gstreamer source frame data, right? You need to append the metadata to the actual frame data.